source/fe/fe_q_base.cc

```diff
@@ -180,12 +180,35 @@ struct FE_Q_Base<PolynomialType, xdim, xspacedim>::Implementation
 
     // use that the element evaluates to 1 at index 0 and along the line at
     // zero
-    const std::vector<unsigned int> &index_map_inverse =
-      fe.poly_space.get_numbering_inverse();
+    const std::vector<unsigned int> &index_map_inverse = [&fe]() {
+      auto *tensor_poly_space_ptr =
+        dynamic_cast<TensorProductPolynomials<dim> *>(fe.poly_space.get());
+      if (tensor_poly_space_ptr != nullptr)
+        return tensor_poly_space_ptr->get_numbering_inverse();
+
+      auto *tensor_piecewise_poly_space_ptr = dynamic_cast<
+        TensorProductPolynomials<dim, Polynomials::PiecewisePolynomial<double>>
+          *>(fe.poly_space.get());
+      if (tensor_piecewise_poly_space_ptr != nullptr)
+        return tensor_piecewise_poly_space_ptr->get_numbering_inverse();
+
+      auto *tensor_bubbles_poly_space_ptr =
+        dynamic_cast<TensorProductPolynomialsBubbles<dim> *>(
+          fe.poly_space.get());
+      if (tensor_bubbles_poly_space_ptr != nullptr)
+        return tensor_bubbles_poly_space_ptr->get_numbering_inverse();
+
+      auto *tensor_const_poly_space_ptr =
+        dynamic_cast<TensorProductPolynomialsConst<dim> *>(fe.poly_space.get());
+      if (tensor_const_poly_space_ptr != nullptr)
+        return tensor_const_poly_space_ptr->get_numbering_inverse();
```

Comment:
This is not pretty, but given this consideration

> There are several instances in `FE_Q_Base` where it assumes the `PolynomialType` class has member functions like `get_numbering()`, `get_numbering_inverse()`, `set_numbering(renumber)`. This would be trouble and unnecessary to try to abstractly implement in `ScalarPolynomialsBase` since it's only used by four of the derived classes and two of them return a reference, so I'm planning to avoid implementing that in `ScalarPolynomialsBase`.

with which I agree, I guess we should find another solution when we have more time: I think the renumbering should really not be in the polynomial classes at all, not even `TensorProductPolynomials`, but rather in `FE_Poly` or `FE_Q_Base`, depending a bit on how we use them. The hierarchic numbering is really a property of the finite element class that is used to impose continuity and conform to how the unknowns should be treated in the finite element classes. Anyway, this is not for this release - if others agree as well, we should open an issue to not forget about that.

Reply:
I like that idea! I think opening an issue to keep track of it is a great idea.

Reply:
@GrahamBenHarper -- can you open such an issue that tracks this?